specification/sdk-tracing.md

```diff
@@ -260,6 +260,53 @@ high contention in a very high traffic service.
 * `maxExportBatchSize` - the maximum batch size of every export. It must be
   smaller or equal to `maxQueueSize`. The default value is `512`.
 
+##### Span watcher processor
+
+This implementation passes spans that have been started but not yet ended
+to the configured exporter. These spans are called _unfinished_ spans.
+On each interval defined by `reportIntervalMillis`, all unfinished spans that
+have been started at least `reportIntervalMillis` ago, are exported.
+
+**Configurable parameters:**
+
+* `exporter` - the exporter to which the spans are pushed.
+* `maxWatchedSpans` - the maximum number of unfinished spans to be watched.
+  While reached, no new spans will be accepted. The default value is `2048`.
+* `reportIntervalMillis` - the mimimum delay interval in milliseconds between two
+  consecutive exports. The default value is `5000`.
+* `exporterTimeoutMillis` - how long the export can run before it is cancelled.
+  The default value is `30000`.
+* `maxExportBatchSize` - the maximum batch size of every export. It must be
+  smaller or equal to `maxWatchedSpans`. The default value is `512`.
+* `exportEndedSpans` (boolean) - if set, spans are also exported once they are
+  ended.
+  The default value is `false`.
+* `maxSpanDurationMillis` - On each report interval, spans older than
+  `maxSpanDurationMillis` which have not yet ended will be dropped from
+  the watch list.
+  The default is `60000` (1 minute).
+  (_only applies to languages where weak references are not supported or
+  suitable_)
+
+In order to avoid multiple exports potentially being blocked on connection issues or
+when `reportIntervalMillis` is set inappropriately, exports SHOULD generally *not*
+be performed in parallel.
+If an export would normally happen while another export by this SpanProcessor is
+still running, the export is skipped.
+
+For languages that support the concept of weak references it is recommended to
+only use these for keeping references to unfinished spans in the processor.
+Otherwise, it would accumulate references to spans abandoned by the user without
+ending them and therefore potentially leak memory.
+If weak references are not supported or suitable to be used, the timeout parameter
+`maxSpanDurationMillis` should be added for this processor. Unfinished spans which
+were first provided to the processor longer than this timeout ago are dropped from the
+list of watched spans and therefore updates will no longer be reported for them.
+This timeout, however, should only apply to this processor and MUST NOT affect spans
+that are ended properly after the timeout. If `exportEndedSpans` is set, the span MUST
```

Comment:
Oof, I need to update my implementation 😅  But makes sense. Should they also be exported if the span watchlist is full at the time they are ended? If yes, is there a dedicated limit for number of queued ended spans?

Reply:
I think both would be okay. It could either be implemented as two separate lists with two limits or one list with a common limit or again two limits. If using one list is easier/more efficient, having two limits would have the advantage that one can set maxEnded > maxWatched so that too many watched spans never suppress sending ended spans in a high load scenario.

Reply:
This is actually a quite interesting question; how to mix ended and unfinished spans. I can see several possibilities:

(NOTE: If you use weak references, then you need to make sure you hold on to the spans in onEnd, which may be implemented by storing spans in a WeakMap instead of a simple List<WeakRef> and setting the value of WeakMap.get(endedSpan). This is an important consideration to estimate the ease of implementation of the variants below.).

1. Don't deal with ended spans at all in the SpanWatcher. Discard a span as soon as it is ended. Employ a separate BatchSpanProcessor for ended spans. HTTP frameworks hopefully deal with recycling the TCP connection, but separate HTTP requests will be made for unfinished and ended spans.
2. Deal with them in the SpanWatcher (as currently specified). Several variants:
   1. Simply send ended spans once, then remove them. If you have strong refs to spans anyway, you don't even need onEnd at all (see note above).
   2. As above, but with "ended span recovery". Consider the case that the watchlist was full at the time a span was started or the span has already timed out (older than maxAge). Then you would still get an onEnd callback in your span processor, which you can use to still potentially send the span. Again, several variants exists:
      1. Single list for ended and unfinished spans: Do in onEnd as you would in onStart, i.e. try to insert the span into the watchlist again. This requires being able to find the span in the watchlist / determining if it is there.
         1. Use the same limit for ended and unfinished spans.
         2. Use a separate limit (typically higher) that allows ended spans to be inserted into the list even if no unfinished would be watched anymore (i.e., reserverve part of the list for only ended spans).
       2. Separate list for ended and unfinished spans. This will require some way to avoid sending spans twice and seeing that spans can be ended at any point in time, might be difficult to do free of race-conditions. But I haven't thought this through.

Reply:
I think I would prefer a 2.ii.a variant, but if someone can describe how to nicely implement 2.ii.b, that would also be fine.

Reply:
Thanks for the detailed write-up! Variant 2.ii.a.b sounds reasonable to me but either way I don't think that the spec should dictate the implementation in that detail. I will define two separate limits for the reason stated in 2.ii.a.b and leave the rest up to the SIGs.

Reply:
I would not overcomplicate this. Once we reach a state that exporter is unable to export fast enough spans are lost and trace is damaged as references are missing. The behavior during such phases is mostly here to avoid more damage to the process. Trying to find a solution to get the best out of the remaining spans is most likely quite useless as the traces with missing spans are garbage.